### PR TITLE
Embedding extraction

### PR DESCRIPTION
**Pull request by setzer22**
Implements #56.

I ported the llama.cpp code to allow extracting word embeddings and logits from a call to `evaluate`. I validated this using an `ad_hoc_test` (currently hard-coded in `main`) and results seem to make sense: The dot product for different embeddings is higher, the more similar the two words are, which is exactly how embeddings should work.

This serves as a proof of concept, but we need to discuss the API before we can merge. Currently, I added an `EvaluateOutputRequest` struct, so we can expand this in the future, allowing retrieval for other interesting bits of the inference process, but these values are not easily obtainable using the regular APIs (i.e. `feed_prompt`, `infer_next_token`). I'm not sure if that's a problem: Are we ok with users having to drop down to the lower level `evaluate` function when they need to retrieve this kind of information?

On a different note, I would really like for someone with a bit of understanding to validate that the results here are correct. Perhaps @hlhr202 can shed some light there?

Finally, should we consider exposing this to `llama-cli` at all?

### CODE REVIEW

File: llama-cli/src/main.rs [hunk 1]
```diff
@@ -314,3 +317,81 @@ fn main() {
         }
     }
 }
+
+fn ad_hoc_test() {
```
Comment (setzer22) [inline llama-cli/src/main.rs:321]: This function will be removed before we merge. It's just here so others can validate my results and test things.

File: llama-cli/src/main.rs [hunk 2]
```diff
@@ -314,3 +317,81 @@ fn main() {
...
+        "I have just adopted a cute ",
+        |_| Ok(()),
+    );
+    let mut output_request2 = EvaluateOutputRequest {
+        all_logits: None,
+        embeddings: Some(Vec::new()),
+    };
+
+    // Try other words: 'dog', 'cat', 'potato', '$' -> To see decreasingly lower dot product values.
+    let dog2 = model.tokenize(&vocab, "dog", false).unwrap();
```
Comment (setzer22) [inline llama-cli/src/main.rs:379]: What I'm doing here is feeding the following two sentences through the transformer:

- "My favourite animal is the dog"
- "I just adopted a cute dog"

Afterwards, I retrieve the embeddings for the last token (dog), and compute their similarity with a simple dot product.

Then, I tried changing the second sentence from 'dog' to 'cat', 'potato', '$' respectively, and the semantic similarity dropped accordingly, with $ ranking the lowest.

File: llama-rs/src/lib.rs [hunk 3]
```diff
@@ -1065,11 +1077,15 @@ impl Model {
     }
 
     /// Evaluates the transformer.
+    ///
+    /// The provided `output_request` struct lets you specify which additional
+    /// data you are interested in fetching from the transformer. Setting a field to a `Some` value will fill the
```
Comment (KerfuffleV2) [inline llama-rs/src/lib.rs:1082]: I think you might have run out of context tokens here. :(

File: llama-cli/src/main.rs (hunk 2 above)
Comment (hlhr202) [reply llama-cli/src/main.rs:379]: @setzer22 will feed prmpt before eval has different embeddings compared to eval all tokens together?

File: llama-rs/src/lib.rs [hunk 4]
```diff
@@ -1296,6 +1316,30 @@ impl Model {
...
 
+        // Extract logits
+        if let Some(all_logits) = &mut output_request.all_logits {
+            all_logits.resize(n_vocab as usize * n, 0.0);
+            // SAFETY: Data can be read (properly aligned, initialized, data
+            // will not be mutated or otherwise aliased while the slice lives),
+            // and we're not reading past the end of the slice.
+            assert_eq!(input_layer.nelements(), n_vocab * n as i32);
+            let logits_slice = unsafe {
+                slice::from_raw_parts(input_layer.data() as *const f32, n_vocab as usize * n)
```
Comment (philpax) [inline llama-rs/src/lib.rs:1327]: Could you add a function to `Tensor` for getting the data out as a slice? This is what I did:
https://github.com/philpax/ggllama/blob/main/src/ggml.rs#L105-L114

**Comment by philpax** [comment]
LGTM once the other review feedback's sorted out. 

For exposing it from the CLI, I'm not sure... people *might* use it as a process in a CLI pipeline (get the embeddings of two texts and then comparing them), but I'm not sure what that would look like or how people would do that. (What output format would we use?)

Unless someone can suggest a "standard" output format for this, I'd suggest leaving it out for now and figuring it out later.

**Comment by hlhr202** [comment]
> LGTM once the other review feedback's sorted out.
> 
> For exposing it from the CLI, I'm not sure... people _might_ use it as a process in a CLI pipeline (get the embeddings of two texts and then comparing them), but I'm not sure what that would look like or how people would do that. (What output format would we use?)
> 
> Unless someone can suggest a "standard" output format for this, I'd suggest leaving it out for now and figuring it out later.

It would be just nice for me to expose such get embedding function as in crate library. actually i do not care much about cli exposing. what I v seen llama.cpp they provide a parameter --embedding for output purpose. but they still did not find out a way to expose it though. thats why i still cannot get the embedding from their cli currently.
I only test a few cases with comparing to openai's embedding. should be some difference, but i think that is caused by different model.

**Comment by KerfuffleV2** [comment]
@hlhr202 The CLI is just a consumer of the library crate, so when using the library you'll be able to get the embeddings.

**Comment by hlhr202** [comment]
> @hlhr202 The CLI is just a consumer of the library crate, so when using the library you'll be able to get the embeddings.

yes absolutely, cuz i m porting llama-rs to llama-node, so i just need library pub function exposing.
it doesnt make sense to expose embeddings in cli anyway.

File: llama-rs/src/lib.rs (hunk 4 above)
Comment (setzer22) [reply llama-rs/src/lib.rs:1327]: Good point!

But, there's a caveat: I don't think this function can be marked as safe. Once something is in a slice, it has to adhere to Rust's aliasing model. The way tensors are implemented, a tensor has to behave like a cheap-to-copy pointer that supports aliasing (this is used in several places of the inference code). This is why tensors have a `share()` function to make this aliasing explicit.

So even if your function tie the lifetime of the tensor to the output slice, there could be many other aliased references to that same tensor through which you could obtain another mutable slice to the same memory. And that's unsound.

What we can do, is have a similar API but mark it as `unsafe` and document the safety requirements for it. But I honestly think it todesn't make sense to try to build an idiomatic safe rust wrapper to ggml. It is a C library built around the idea of aliasing pointers and "being careful" not to drop a context before some other context uses it :sweat_smile:. Even if Rust lifetimes could somehow capture that, the loss in ergonomics would be a high price to pay. One of the main design goals in the original design was to make the code as easy to port as possible. This includes any changes to the tensor code that may happen in the future, like #68.

Any efforts towards making ggml more Rusty would be better directed towards #31.

File: llama-cli/src/main.rs (hunk 2 above)
Comment (setzer22) [reply llama-cli/src/main.rs:379]: @hlhr202 The embeddings wouldn't be affected, but you shouldn't call `evaluate` with the whole prompt like that for a couple reasons:

- A call to evaluate runs all the tokens you give it as a batch, meaning it requires increased memory usage. For very long prompts, this could become very expensive.
- The output value will return the output embeddings for every token that you fed through `eval`. This means you would be retrieving a lot more embedding data than for just the word "dog".

File: llama-rs/src/lib.rs (hunk 4 above)
Comment (setzer22) [reply llama-rs/src/lib.rs]: I just realized we already had a similar function for this (`read_data`), so I used that one instead :+1:

**Comment by setzer22** [comment]
I already addressed the review feedback and removed the ad-hoc test code. So I take it a good plan now would be to merge this as-is and have embedding extraction as a low-level feature of llama-rs, but simply not expose it to the CLI?

File: llama-rs/src/lib.rs (hunk 3 above)
Comment (philpax) [reply llama-rs/src/lib.rs:1082]: @setzer22 fyi

**Review by philpax — APPROVED** [review]
LGTM, ready to merge after the comment's fixed

**Comment by KerfuffleV2** [comment]
Since I added the `--dump-prompt-tokens` option, you can probably guess I like exposing information. :) I know people asked about being able to show the embeddings with llama.cli, so it does seem like there's some kind of demand for it in a CLI.

**Comment by philpax** [comment]
If there's demand, I'm happy to do so - just not sure what the output format should be. JSON array or newline-delimited floats?

**Comment by KerfuffleV2** [comment]
Is it a lot of data? You could probably just print in the normal Rust debug format which should look like a comma separated list if it's in a `Vec` or something. That should be pretty easy to transform to other formats without need to write extra code or pull in dependencies.

File: llama-rs/src/lib.rs (hunk 3 above)
Comment (setzer22) [reply llama-rs/src/lib.rs]: Oops, missed that one, should be fixed now :+1:

**Comment by setzer22** [comment]
> Is it a lot of data?

It is quite a lot of data for comfortably printing to stdout honestly. It's 4096 floats per token.

**Comment by KerfuffleV2** [comment]
Ahh, then seems like it probably isn't worth even bothering to add to the CLI right now unless someone comes here and requests it. Or they could probably just write their own little utility to load a model, feed a prompt and print out embeddings however they wanted.

File: llama-cli/src/main.rs (hunk 2 above)
Comment (hlhr202) [reply llama-cli/src/main.rs]: @setzer22 I just understand your comments here. this means we can only extract embeddings for a single part of words (which may also have hidden information mixed with context of the whole sentence). that should a little bit different with OpenAI's embedding function. what i understand for openai's embedding, is for the whole sentence but at the same time returned in a fixed size of tensor... that is quite beyond my knowledge though

Comment (hlhr202) [reply llama-cli/src/main.rs]: well I guess I might find possible ways to implement such 'sentence embedding', I will try add some special end token and extract the hidden layer once the end token evaluated. not sure if it works, but it must worth a try.

**Comment by rpbrokaw** [comment]
> Ahh, then seems like it probably isn't worth even bothering to add to the CLI right now unless someone comes here and requests it. Or they could probably just write their own little utility to load a model, feed a prompt and print out embeddings however they wanted.

I would love that in the CLI! Perhaps with a parameter that specifies an output file. I need the embeddings to build a vector database based on some local files. Any chance you could take a look? It has been many years since I programmed C/C++.

**Comment by hlhr202** [comment]
> > Ahh, then seems like it probably isn't worth even bothering to add to the CLI right now unless someone comes here and requests it. Or they could probably just write their own little utility to load a model, feed a prompt and print out embeddings however they wanted.
> 
> I would love that in the CLI! Perhaps with a parameter that specifies an output file. I need the embeddings to build a vector database based on some local files. Any chance you could take a look? It has been many years since I programmed C/C++.

vector is around 4096 length for one token, not very well printed in CLI. I guess you need to call in through rust api.

**Comment by setzer22** [comment]
I'm open to adding a way for the CLI to output embeddings if people find this is an interesting use case. The main blocker here is that the use case is not clear to me and thus I can't figure out the right API and output format.

What we need here, is someone who understands how embeddings in an LLM like LLaMA work, and has a clear use case for extracting them and can tell us how would they expect an API like this to work. If anyone wants to open an issue with a clear description of what we need to provide, I'd be happy to add an implementation :slightly_smiling_face:

**Comment by hlhr202** [comment]
@setzer22 I have made a new embedding extraction example. can check it here https://github.com/hlhr202/llama-node/blob/develop/packages/core/example/semantic-compare/compare.py
I noticed that llama.cpp they use "\n" as end token, so I also do the same. it is quite close to openai's text-embedding-ada-002 result

**Comment by turbo** [comment]
I'm working on a large dense-vector embedding database (about 2 million data points from books), which is currently using OpenAI's Ada embeddings (~1600 dimensions). I can do a comparison of performance between those and the 4k LLaMa embeds if needed.

> as a clear use case for extracting them and can tell us how would they expect an API like this to work

From an ops perspective, ideally one could provide a batch input and get a batch output (just like OpenAI's API) via CLI. The format doesn't matter much - it can be JSONL or a binary format. I'd personally recommend sticking to those two, since that is supported by most VSS databases (e.g. Redis RediSearch).

**Comment by merlinvn** [comment]
> o adding a way for the CLI to output embeddings if pe
My use case here is if you have a sets of documents, and if you can get the embeddings of those documents, whenever a new question comming in, you can embed the new questions and find the most relevant documents to send a long with you prompt. So basically you can have a natural Q&A chat bot based on your own data.
